Mathlib/AlgebraicGeometry/RationalMap.lean

```diff
@@ -0,0 +1,308 @@
+  ∃ (W : X.Opens) (hW : Dense (W : Set X)) (hWl : W ≤ f.domain) (hWr : W ≤ g.domain),
+    (f.restrict W hW hWl).hom = (g.restrict W hW hWr).hom
+
+lemma iseqv_rel : _root_.Equivalence (@Scheme.PartialMap.equiv X Y) where
+  refl f := ⟨f.domain, f.dense_domain, by simp⟩
+  symm {f g} := by
+    intro ⟨W, hW, hWl, hWr, e⟩
+    exact ⟨W, hW, hWr, hWl, e.symm⟩
+  trans {f g h} := by
+    intro ⟨W₁, hW₁, hW₁l, hW₁r, e₁⟩ ⟨W₂, hW₂, hW₂l, hW₂r, e₂⟩
+    refine ⟨W₁ ⊓ W₂, hW₁.inter_of_isOpen_left hW₂ W₁.2, inf_le_left.trans hW₁l,
+      inf_le_right.trans hW₂r, ?_⟩
+    dsimp at e₁ e₂
+    simp only [restrict_domain, restrict_hom, restrictFunctor_obj_left, homOfLE_leOfHom,
+      ← X.homOfLE_homOfLE (U := W₁ ⊓ W₂) inf_le_left hW₁l, Functor.map_comp, Over.comp_left,
+      Category.assoc, e₁, ← X.homOfLE_homOfLE (U := W₁ ⊓ W₂) inf_le_right hW₂r, ← e₂]
+    simp only [homOfLE_homOfLE_assoc]
+
+instance : Setoid (X.PartialMap Y) := ⟨@PartialMap.equiv X Y, iseqv_rel⟩
+
+lemma equiv_of_fromSpecStalkOfMem_eq [IrreducibleSpace X]
+    {x : X} [X.IsGermInjectiveAt x] (f g : X.PartialMap Y)
+    (hxf : x ∈ f.domain) (hxg : x ∈ g.domain)
+    (H : f.fromSpecStalkOfMem hxf = g.fromSpecStalkOfMem hxg) : f.equiv g := by
+  have hdense : Dense ((f.domain ⊓ g.domain) : Set X) :=
+    f.dense_domain.inter_of_isOpen_left g.dense_domain f.domain.2
+  have := (isGermInjectiveAt_iff_of_isOpenImmersion (f := (f.domain ⊓ g.domain).ι)
+    (x := ⟨x, hxf, hxg⟩)).mp ‹_›
+  have := spread_out_unique_of_isGermInjective' (X := (f.domain ⊓ g.domain).toScheme)
+    (X.homOfLE inf_le_left ≫ f.hom) (X.homOfLE inf_le_right ≫ g.hom) (x := ⟨x, hxf, hxg⟩) ?_
+  · obtain ⟨U, hxU, e⟩ := this
+    refine ⟨(f.domain ⊓ g.domain).ι ''ᵁ U, ((f.domain ⊓ g.domain).ι ''ᵁ U).2.dense
+      ⟨_, ⟨_, hxU, rfl⟩⟩,
+      ((Set.image_subset_range _ _).trans_eq (Subtype.range_val)).trans inf_le_left,
+      ((Set.image_subset_range _ _).trans_eq (Subtype.range_val)).trans inf_le_right, ?_⟩
+    rw [← cancel_epi (Scheme.Hom.isoImage _ _).hom]
+    simp only [TopologicalSpace.Opens.carrier_eq_coe, IsOpenMap.functor_obj_coe,
+      TopologicalSpace.Opens.coe_inf, restrict_hom, ← Category.assoc] at e ⊢
+    convert e using 2 <;> rw [← cancel_mono (Scheme.Opens.ι _)] <;> simp
+  · rw [← f.fromSpecStalkOfMem_restrict _ hdense inf_le_left ⟨hxf, hxg⟩,
+      ← g.fromSpecStalkOfMem_restrict _ hdense inf_le_right ⟨hxf, hxg⟩] at H
+    simpa only [fromSpecStalkOfMem, restrict_domain, Opens.fromSpecStalkOfMem, Spec.map_inv,
+      restrict_hom, Category.assoc, IsIso.eq_inv_comp, IsIso.hom_inv_id_assoc] using H
+
+end PartialMap
+
+/-- A rational map from `X` to `Y` (`X ⤏ Y`) is an equivalence class of partial maps,
+where two partial maps are equivalent if they are equal on a dense open subscheme.  -/
+def RationalMap (X Y : Scheme.{u}) : Type u :=
+  @_root_.Quotient (X.PartialMap Y) inferInstance
+
+/-- The notation for rational maps. -/
+scoped[AlgebraicGeometry] infix:10 " ⤏ " => Scheme.RationalMap
```

Comment:
Is this already typeable in VS code? (It is not in neovim)

Reply:
No. I will add a PR later.